### DISCUSSION TITLE
how to do live update in plotly in nicegui?

### DISCUSSION THREAD

**Discussion by future** [zauberzeug/nicegui#636] [Q&A] [answered]
May I ask how to do live update in plotly in nicegui?
How to dynamically update a graph in plotly?

Thanks

**Comment by falkoschindler** [comment, marked as answer]
Good question, @future!

Similar to the [demo about UI updates](https://nicegui.io/documentation#ui_updates), a `ui.plotly` element needs an explicit `update()` call to synchronize with the client. Here is an example:

```py
fig = go.Figure()
fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
plot = ui.plotly(fig).classes('w-full h-40')

def add_trace():
    fig.add_trace(go.Scatter(x=np.arange(10), y=np.random.randn(10)))
    plot.update()

ui.button('Add trace', on_click=add_trace)
```

I'll add this demo to the [documentation](https://nicegui.io/documentation/plotly).

**Reply by future** [reply]
Thank you very much for your answer, I tried to imitate the "Line Plot" paradigm to dynamically update the data in the chart, but found that clear() seems to be unable to clear, I wonder if I misunderstood something?

thank you very much

Below is my test code:

```py
import numpy as np
from nicegui import ui
import plotly.graph_objects as go
from nicegui import ui

t = 5
x_data = np.arange(0, 5)
y_data = x_data*2

fig = go.Figure()
fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
plot = ui.plotly(fig).classes('w-full h-40')

def add_trace():
    fig.add_trace(go.Scatter(x=x_data, y=y_data))
    plot.update()

def clear_trace():
    plot.clear()

def update_data():
    global t
    t += 1

    x_data = np.arange(0, t)
    y_data = x_data*2

    plot.clear()
    fig.add_trace(go.Scatter(x=x_data, y=y_data))
    plot.update()

line_updates = ui.timer(0.5, update_data, active=False)

with ui.row():
    ui.button('Add trace', on_click=add_trace)
    ui.button('Clear trace', on_click=clear_trace)
    line_checkbox = ui.checkbox('active').bind_value(line_updates, 'active')

ui.run()
```

**Reply by falkoschindler** [reply]
Yeah, the `clear()` method [keeps confusing](https://github.com/zauberzeug/nicegui/issues/414) people. It is inherited from `ui.element` and has nothing to do with Plotly. As its docstring "Remove all child elements." says, it is intended to be used for removing nested NiceGUI elements like labels and buttons in a container like a row or a column. Maybe we should overwrite this behavior for `ui.plotly` like [we did for `ui.log`](https://github.com/zauberzeug/nicegui/discussions/516).

Anyway, to clear all traces, you can set `data = []`:

```py
def clear_trace():
    fig.data = []
    plot.update()
```

https://community.plotly.com/t/remove-all-traces/13469

**Reply by future** [reply]
Thank you very much for your detailed answer, I wish NiceGUI better and better !!
